tests/components/mqtt/test_init.py

```diff
@@ -2046,6 +2046,26 @@ async def test_logs_error_if_no_connect_broker(
     )
 
 
+@pytest.mark.parametrize("return_code", [4, 5])
+async def test_triggers_reauth_flow_if_auth_fails(
+    hass: HomeAssistant,
+    caplog: pytest.LogCaptureFixture,
+    mqtt_mock_entry: MqttMockHAClientGenerator,
+    mqtt_client_mock: MqttMockPahoClient,
+    return_code: int,
+) -> None:
+    """Test re-auth is triggered if authentication is failing."""
+    await mqtt_mock_entry()
+    config_entry = hass.config_entries.async_entries(mqtt.DOMAIN)[0]
+    # test with rc = 4 -> CONNACK_REFUSED_NOT_AUTHORIZED and 5 -> CONNACK_REFUSED_BAD_USERNAME_PASSWORD
+    with patch.object(
+        config_entry, "async_start_reauth", MagicMock()
+    ) as mock_async_start_reauth:
+        mqtt_client_mock.on_connect(mqtt_client_mock, None, None, return_code)
+        mock_async_start_reauth.assert_called_once()
+    await hass.async_block_till_done()
```

Comment:
It would be cleaner to check for active flows vs patching `async_start_reauth`

```
    flows = hass.config_entries.flow.async_progress()
    assert len(flows) == 1
    assert flows[0]["context"]["source"] == "reauth"
```